proj/hps_accel/gateware/get.py

```diff
@@ -32,20 +40,32 @@ class StatusRegister(SimpleElaboratable):
       A sink for new values. "ready" is always asserted because the
       register is always ready to receive new data.
 
-    value: unsigned(32), out
+    invalidate: Signal(), in
+      Causes valid to be deassserted.
+
+    value: Signal(32), out
       The value held by the register. Received from sink.payload.
 
+    valid: Signal(), out
```

Comment:
This is interesting. I ran into this issue last night (lack of flow control between the store and the get registers) but the first solution which came to mind was to make the get registers act more like a proper stream consumer -- when the CPU accesses the register, its sink becomes "ready" and the next value is consumed. The CPU would block until the connected source is "valid".

But that would mean that every value sent to the register must be read exactly once. If the CPU doesn't read it, or wants to read it more than once, everything deadlocks. So it's a fragile design.

Whereas signalling validity of the register value means that the CPU is free to read out the register multiple times, as long as the value is valid, which is more like the traditional behaviour of a CPU register. But then in the doc string you note that each register can only be read a single time?

Reply:
Oops. The design process leaked into the comments. Yes, these were exactly the trade-offs I was making and eventually settled on read-many for now.

It would be possible, without too much effort to make some registers read-once, and I think we will need to do that.

Reply:
On re-reading the code and my reply, perhaps I was a little confused. Here is the current state: 

- registers can be set by the gateware many times. They are always ready to receive a new value
- registers can be read by the CPU many times. However, if they are not "valid", then the read will block. This is a different "valid" than the register's sink has.
- valid stays asserted unless:
  - valid_at_reset is false, then valid will be deasserted at reset
  - invalidate is asserted.
- after a value is received from the sink, then valid is asserted.

Reply:
Fixed the wrong comment in constants.py.